### DISCUSSION TITLE
Feature Request: Execution/Query Progress API for Python

### DISCUSSION THREAD

**Discussion by kevinschaich** [duckdb/duckdb#11923] [Ideas]
## Why do you want this feature?

A lot of people have expressed interest in being able to monitor the progress of long-running queries as they're happening.

## Current State

- A progress bar is currently available for the CLI ([`PRAGMA enable_progress_bar;`](https://duckdb.org/docs/configuration/pragmas#progress-bar))
- A function to get progress of the running query is available for the C client API ([`duckdb_query_progress`](https://duckdb.org/docs/api/c/api#duckdb_query_progress))
- There's not currently a way to programatically get progress of a query in Python (AFAIK)

### Option 1: Add an option to provide a callback/handler for query progress to the DuckDB connection object

This is how it works in @Tishj 's PR. I propose also adding more details to this callback, such as the current stage that the query is on, number of rows read in that stage, scan speed, etc – would be useful to show _what_ is happening in addition to how far along the total progress is.

Example:

```python
def progress_handler(state, progress, finished):
    print(progress)

con = duckdb.connect()
con.set_progress_handler(progress_handler)
res = con.execute("SELECT * FROM my_table")
```

### Option 2: Asynchronous execution cursor pointer that supports progress

Polling for updates on a pointer to an eventual result (related: [@motherduckdb/duckdb-async](https://github.com/motherduckdb/duckdb-async)):

```python
con = duckdb.connect()
res = con.execute_async("SELECT * FROM my_table")

while (res.result is None):
    print(res.progress)
    time.sleep(1)

print(res.result)
```

### Option 3: Pipe DuckDB output into stdout 

I've done this with other packages but I couldn't find a way for DuckDB:

```python
import duckdb
import logging
print(list(logging.root.manager.loggerDict.keys())) # nothing related to Duckdb
```

I also tried [what others have mentioned working in Jupyter](https://github.com/duckdb/duckdb/issues/6815) (first enabling the progress bar, then printing the `execute` statement) but at least for me this is synchronous and doesn't print until the query is finished

```python
con = duckdb.connect(database='test.db', read_only=False)
con.sql("SET enable_progress_bar = TRUE;")
con.sql("SET progress_bar_time	= 100;")
print(con.sql("SELECT * FROM 'http://blobs.duckdb.org/train_services.parquet' LIMIT 1;"))
```

## Related Issues & Discussions

#983 (query-specific progress bars [only in CLI])
#6815 (kernel crashes with progress bar in jupyterlab)
#7590 (ipywidgets is required to set progress bar setting)
#9066 (API to get the progress bar)
#7100 (Get progress bar for a Python GUI)

### PR Implementation ([Tishj/duckdb](https://github.com/Tishj/duckdb) fork)

https://github.com/Tishj/duckdb/pull/26

---

cc: @Tishj @ben-pr-p @davetapley

**Comment by joshuasamuelevans** [comment]
I would just like to add to this thread that this feature would be an amazing utility to have, especially when running large spatial queries.

**Comment by ben-pr-p** [comment]
Option 1 seems to be the best to me! I am planning on writing progress to a database to update a user waiting screen, so anything using stdout would require some mangling with pipes on my end.

**Comment by nickzoic** [comment]
I've tried a bit of a different approach at #16927 which is just to have a `conn.query_progress()` method which can be run from a different thread (much like `conn.interrupt()`), which works but you've got to have progress bars enabled and also only certain kinds of query give meaningful results.
